linker/linker_test.go

```diff
@@ -1311,3 +1312,67 @@ func TestLinkerSymbolCollisionNoSource(t *testing.T) {
 	require.Error(t, err)
 	assert.EqualError(t, err, `foo.proto: symbol "google.protobuf.DescriptorProto" already defined at google/protobuf/descriptor.proto`)
 }
+
+func TestSyntheticOneOfCollisions(t *testing.T) {
+	t.Parallel()
+	input := map[string]string{
+		"foo1.proto": `
+			syntax = "proto3";
+			message Foo {
+			  optional string bar = 1;
+			}`,
+		"foo2.proto": `
+			syntax = "proto3";
+			message Foo {
+			  optional string bar = 1;
+			}`,
+	}
+
+	var errs []error
+	compiler := &protocompile.Compiler{
+		Reporter: reporter.NewReporter(
+			func(err reporter.ErrorWithPos) error {
+				errs = append(errs, err)
+				// need to return nil to accumulate all errors so we can report synthetic
+				// oneof collision; otherwise, the link will fail after the first collision
+				// and we'll never test the synthetic oneofs
+				return nil
+			},
+			nil,
+		),
+		Resolver: protocompile.ResolverFunc(func(filename string) (protocompile.SearchResult, error) {
+			f, ok := input[filename]
+			if !ok {
+				return protocompile.SearchResult{}, fmt.Errorf("file not found: %s", filename)
+			}
+			return protocompile.SearchResult{Source: strings.NewReader(removePrefixIndent(f))}, nil
+		}),
+	}
+	_, err := compiler.Compile(context.Background(), "foo1.proto", "foo2.proto")
+
+	assert.Equal(t, reporter.ErrInvalidSource, err)
+
+	// since files are compiled concurrently, there are two possible outcomes
+	expectedFoo1FirstErrors := []string{
+		`foo2.proto:2:9: symbol "Foo" already defined at foo1.proto:2:9`,
+		`foo2.proto:3:19: symbol "Foo.bar" already defined at foo1.proto:3:19`,
+		`foo2.proto:3:19: symbol "Foo._bar" already defined at foo1.proto:3:19`,
```

Comment:
Error message handling has changed significantly (and looks better in protocompile): https://github.com/jhump/protoreflect/pull/512/files#diff-f213d9afe67d61fc59d87d32853b67d02b40ad569e3c6d48620ccfe646a802dfR1072-R1076

I updated this tests to match existing behavior (which quotes duplicate symbols and also includes source code info of both declarations).